File: built-in plugins dsp/mb-synth-trance.cpp
/**
 * MB Trance Lead
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic trance lead
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SYNTH_TRANCE_H
#define MB_SYNTH_TRANCE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSynthTrance : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-synth-trance";
    static constexpr const char* PLUGIN_NAME    = "MB Trance Lead";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float gate = 0.7f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSynthTrance() = default;
    ~MbSynthTrance() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.gate = std::clamp(params.gate, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Trance Lead
        return input;
    }
};

#endif // MB_SYNTH_TRANCE_H

File: built-in plugins dsp/mb-synth-trance.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSynthTrancePlugin: PluginDefinition = {
    id: 'mb-synth-trance', slug: 'mb-synth-trance', name: 'MB Trance Lead', category: 'instrument', type: 'analog', version: '1.0.0',
    description: 'Classic trance lead', author: 'Max Booster',
    oscillators: [{ type: 'sawtooth', detune: -10, gain: 0.25 }, { type: 'sawtooth', detune: 10, gain: 0.25 }, { type: 'square', detune: 0, gain: 0.5 }],
    envelope: { attack: 0.01, decay: 0.3, sustain: 0.5, release: 0.4 },
    parameters: [
      { id: 'gate', name: 'Gate', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { gate: 0.7, volume: 0.8 },
  };

export default MbSynthTrancePlugin;
